pkg/models/system_intake_grb_discussions.go

```diff
@@ -0,0 +1,91 @@
+package models
+
+import (
+	"errors"
+	"slices"
+
+	"github.com/google/uuid"
+)
+
+type SystemIntakeGRBReviewDiscussionPost struct {
+	BaseStructUser
+	Content        HTML                     `json:"content" db:"content"`
+	SystemIntakeID uuid.UUID                `json:"systemIntakeId" db:"system_intake_id"`
+	ReplyToID      *uuid.UUID               `db:"reply_to_id"`
+	VotingRole     *SIGRBReviewerVotingRole `json:"votingRole" db:"voting_role"`
+	GRBRole        *SIGRBReviewerRole       `json:"grbRole" db:"grb_role"`
+}
+
+func NewSystemIntakeGRBReviewDiscussionPost(createdBy uuid.UUID) *SystemIntakeGRBReviewDiscussionPost {
+	return &SystemIntakeGRBReviewDiscussionPost{
+		BaseStructUser: NewBaseStructUser(createdBy),
+	}
+}
+
+// CreateGRBDiscussionsFromPosts sorts a slice of discussion posts (replies and initial) and sorts them into multiple discussions
+func CreateGRBDiscussionsFromPosts(posts []*SystemIntakeGRBReviewDiscussionPost) ([]*SystemIntakeGRBReviewDiscussion, error) {
+	postMap := map[uuid.UUID][]*SystemIntakeGRBReviewDiscussionPost{}
+	for _, post := range posts {
+		// shouldn't happen but we deference below
+		if post == nil {
+			return nil, errors.New("post is nil")
+		}
+		// group posts into slices by the initial post id
+		var groupingID uuid.UUID
+		if post.ReplyToID != nil {
+			groupingID = *post.ReplyToID
+		} else {
+			groupingID = post.ID
+		}
+		posts, ok := postMap[groupingID]
+		if ok {
+			postMap[groupingID] = append(posts, post)
+		} else {
+			postMap[groupingID] = []*SystemIntakeGRBReviewDiscussionPost{post}
+		}
+	}
+
+	// after grouping by initial post ID, loop through slice of slices and convert groups of posts into discussions
+	discussions := []*SystemIntakeGRBReviewDiscussion{}
+	for _, posts := range postMap {
+		groupedPosts, err := CreateGRBDiscussionFromPosts(posts)
+		if err != nil {
+			return nil, err
+		}
+		discussions = append(discussions, groupedPosts)
+	}
+	slices.SortFunc(discussions, func(a *SystemIntakeGRBReviewDiscussion, b *SystemIntakeGRBReviewDiscussion) int {
+		return a.InitialPost.CreatedAt.Compare(b.InitialPost.CreatedAt)
+	})
+	return discussions, nil
+}
+
+// CreateGRBDiscussionFromPosts organizes a post and its replies into a single discussion
+// (for slices with multiple initial posts/related replies use CreateGRBDiscussionsFromPosts)
+func CreateGRBDiscussionFromPosts(posts []*SystemIntakeGRBReviewDiscussionPost) (*SystemIntakeGRBReviewDiscussion, error) {
+	var discussion SystemIntakeGRBReviewDiscussion
+	for _, post := range posts {
+		if post.ReplyToID == nil {
+			if discussion.InitialPost != nil {
```

Comment:
from what i can see, wouldn't this always be false? `discussion` is declared a few lines up without any of its fields being initialized

Reply:
Since this check is in the `for` loop, I think this code is just checking that we basically don't have more than one `post` with a `ReplyToID` that is `nil` (i.e. a top-level post), since just after this check we set `discussion.InitialPost`.

There's maybe some comments we could add or refactoring to make it a bit more clear, however. Maybe something like this?

```golang
if post.ReplyToID == nil { // if this is true, it's an initial post (not replying to anything)
    if discussion.InitialPost != nil { // if we *already* set discussion.InitialPost, then there's something wrong (2 top level posts?)
```

Reply:
ha! you are right, it's in the loop 🤦 can't believe i missed that